### PR TITLE
Implement storage::internal::CurlDownloadRequest.

### PR DESCRIPTION
**Pull request by coryan**
Part of the fixes for #819. This is a libcurl wrapper to receive
large downloads without blocking or allocating memory to hold the
full download.

### CODE REVIEW

File: google/cloud/storage/internal/curl_download_request.cc [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+  }
+  return running_handles;
+}
+
+void CurlDownloadRequest::WaitForHandles() {
+  CURLMcode result = curl_multi_wait(multi_.get(), nullptr, 0, 0, nullptr);
+  HandleCurlMultiErrorCode(__func__, result);
+}
+
+void CurlDownloadRequest::HandleCurlMultiErrorCode(char const* where,
```
Comment (houglum) [inline google/cloud/storage/internal/curl_download_request.cc]: From reading the name of the function, I'd assume I should only pass failure codes to it (not a return code indicating a success).

How about something like `RaiseErrorIfBadCurlMultiReturnCode`?

File: google/cloud/storage/internal/curl_download_request.h [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    handle_ = std::move(rhs.handle_);
+    multi_ = std::move(rhs.multi_);
+    curl_closed_ = rhs.curl_closed_;
+    initial_buffer_size_ = rhs.initial_buffer_size_;
+    ResetOptions();
+    return *this;
+  }
+
+  /**
+   * Get some more data into @p buffer.
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/internal/curl_download_request.h]: Fixed.

File: google/cloud/storage/internal/curl_download_request.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/internal/curl_download_request.cc]: Why not just `RaiseOnError()`?  I fixed the other function that was also confusingly named. And we might refactor these two classes in the future, but right now I am not sure if they have enough commonality.

**Comment by coryan** [comment]
@houglum PTAL.

File: google/cloud/storage/internal/curl_download_request.cc (hunk 1 above)
Comment (houglum) [reply google/cloud/storage/internal/curl_download_request.cc]: Sounds good.

**Pull request merged**